addons/website/static/src/interactions/state_form_field.js

```diff
@@ -0,0 +1,80 @@
+import { markup } from "@odoo/owl";
+import { Interaction } from "@web/public/interaction";
+import { registry } from "@web/core/registry";
+
+export class StateFormField extends Interaction {
+    static selector =
+        ".s_website_form:has([name='country_id']) .s_website_form_input[data-link-state-to-country='true']";
+    dynamicSelectors = {
+        ...this.dynamicSelectors,
+        _countryField: () =>
+            this.el
+                .closest(".s_website_form")
+                .querySelector(".s_website_form_input[name='country_id']"),
+    };
+    dynamicContent = {
+        _root: {
+            "t-on-change": this.onStateFieldChange,
+            "t-out": this.updateStateOptions,
+        },
+        _countryField: {
+            "t-on-change": this.onCountryFieldChange,
+        },
+    };
+
+    setup() {
+        // Cloning ensures we keep a detached copy
+        this.allOptions = Array.from(this.el.querySelectorAll("option")).map((opt) =>
+            opt.cloneNode(true)
+        );
+
+        this.countryField = this.el
+            .closest("form")
+            .querySelector(".s_website_form_input[name='country_id']");
+        this.selectedCountryId =
+            this.countryField.value || this.countryField.querySelector("[selected]")?.value || "";
+
+        this.availableCountriesIds = Array.from(this.countryField.querySelectorAll("option")).map(
+            (opt) => opt.value
+        );
+    }
+
+    onCountryFieldChange() {
+        this.selectedCountryId = this.countryField.value;
+    }
+
+    onStateFieldChange() {
+        this.selectedStateId = this.el.value;
+        if (!this.selectedStateId) {
+            return;
+        }
+        this.selectedCountryId =
+            Array.from(this.el.options).find((option) => option.value === this.selectedStateId)
+                ?.dataset.countryId || "";
+        this.countryField.value = this.selectedCountryId;
+    }
+
+    updateStateOptions() {
+        const options = this.allOptions.filter((option) => {
+            if (option.value !== this.selectedStateId) {
+                option.removeAttribute("selected");
+            } else {
+                option.setAttribute("selected", "selected");
+            }
+            const optionCountryId = option.getAttribute("data-country-id") || "";
+            return (
+                (!this.selectedCountryId && this.availableCountriesIds.includes(optionCountryId)) ||
+                !option.value ||
+                optionCountryId === this.selectedCountryId
+            );
+        });
+
+        // disable the state field if there is no option to select,
+        // or if there's only the empty one
+        this.el.disabled = options.length <= 1 && !options[0]?.value;
+
+        return markup(options.map((option) => option.outerHTML).join(""));
```

Comment:
Just an idea:
- In the setup, extract a list with the data on the option
- use renderAt with a small template something like <option t-foreach=...> + the filter list of option ?
(We don t need to use markup so we don t need to ping secu :) )